File: src/main/java/ru/openbank/meetup/demo/java/slide4/WhenUseCase.java
package ru.openbank.meetup.demo.java.slide4;

import java.util.regex.Pattern;


public final class WhenUseCase {

    private WhenUseCase() {
    }

    private static final Pattern PATTERN = Pattern.compile("^((?!0{10})\\d{10})|((?!0{12})\\d{12})$");

    public static boolean isValid(String inn) {
        if (inn == null) {
            //возвращаем true, т.к. для проверки на null должна использоваться аннотация @NotNull
            return true;
        }

        if (!PATTERN.matcher(inn).matches()) {
            return false;
        }

        if (inn.length() == 12) {
            return isValidINNWitTwelveNumbers(inn);
        }

        if (inn.length() == 10) {
            return isValidINNWithTenNumbers(inn);
        }

        return false;
    }

    private static boolean isValidINNWithTenNumbers(String sourceInn) {
        return true;

    }

    private static boolean isValidINNWitTwelveNumbers(String sourceInn) {
        return true;
    }
}

File: src/main/kotlin/ru/openbank/meetup/demo/kotlin/slide4/WhenUseCase.kt
package ru.openbank.meetup.demo.kotlin.slide4

import java.util.regex.Pattern

@Suppress("MagicNumber")
object InnValidator {

    private val PATTERN = Pattern.compile("^((?!0{10})\\d{10})|((?!0{12})\\d{12})$")

    private const val INN_OF_INDIVIDUAL_LENGTH = 12
    private const val INN_OF_LLC_LENGTH = 10

    fun isValid(inn: String?): Boolean {
        return when (inn?.length) {
            null -> true
            INN_OF_LLC_LENGTH -> inn.isCorrectInn() && isValidINNWithTenNumbers(inn)
            INN_OF_INDIVIDUAL_LENGTH -> inn.isCorrectInn() && isValidINNWitTwelveNumbers(inn)
            else -> false
        }
    }

    private fun String.isCorrectInn(): Boolean = PATTERN.matcher(this).matches()

    private fun isValidINNWithTenNumbers(sourceInn: String): Boolean {
        return true;
    }

    private fun isValidINNWitTwelveNumbers(sourceInn: String): Boolean {
        return true
    }
}

fun main() {
    InnValidator.isValid("")
}
